Lib/tkinter/filedialog.py

```diff
@@ -385,9 +385,27 @@ def askopenfilename(**options):
 
 
 def asksaveasfilename(**options):
-    "Ask for a filename to save as"
+    """Ask for a filename to save as"""
+
+    # If no default extension is provided, set it to the first filetype
+    if "defaultextension" not in options:
+        filetypes = options.get("filetypes")
+        if filetypes:
+            first_ext = filetypes[0][1]  # Get "*.txt" from ("Text files", "*.txt")
+            if first_ext.startswith("*"):
+                options["defaultextension"] = first_ext[1:]  # ".txt"
+
+    filename = SaveAs(**options).show()
+
+    # Append extension if missing
+    if filename and '.' not in filename:
+        ext = options.get("defaultextension", "")
+        if ext and not filename.endswith(ext):
+            filename += ext
```

Comment:
This will not work for "hidden" dot-files, for file names containing a dot in the middle by accident, for relative file paths containing "../" or "./". It is easier to add an extension in your program after returning from `asksaveasfilename()`, in a way that specific to your program.